lbrynet/blob/blob_file.py

```diff
@@ -16,19 +16,21 @@
 blobhash_length = get_lbry_hash_obj().digest_size * 2
 
 
-def is_valid_hashcharacter(char):
-    return char in "0123456789abcdef"
-
+def is_hex_character(char):
+    try:
+        return 16 > int(char, 16) >= 0
+    except ValueError:  # Raised if non base16 string given
+        return False
 
 def is_valid_blobhash(blobhash):
     """Checks whether the blobhash is the correct length and contains only
-    valid characters (0-9, a-f)
+    hexadecimal characters: (0-9, a-f).
 
     @param blobhash: string, the blobhash to check
 
     @return: True/False
     """
-    return len(blobhash) == blobhash_length and all(is_valid_hashcharacter(l) for l in blobhash)
+    return len(blobhash) == blobhash_length and all(is_hex_character(l) for l in blobhash)
```

Comment:
Using `unhexlify` would be about 5x faster than manually looping over individual characters and calling the `is_hex_character` function.
```
Python 3.7.0 (default, Aug  6 2018, 20:07:46) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import timeit
>>> timeit.timeit('try:\n  unhexlify("beef")\nexcept Error:\n  pass', 'from binascii import unhexlify', number=10000)
0.0037124441005289555
>>> timeit.timeit('all(is_hex(c) for c in "beef")', 'def is_hex(c): return c in "0123456789abcdef"', number=10000)
0.024959808913990855
```

Reply:
In this case the entire line should just be replaced with the `unhexlify` instead of what was in the commit prior to mine.

Reply:
Sure, only catch is that `unhexlify()` requires `bytes` only and fails on if you pass it a `str` so changing the function to use `unhexlify()` will require careful testing to make sure any code using `is_valid_blobhash` only passes `bytes` and not `str`.

Reply:
I looked into it and `unhexlify()` wouldn't be suitable since it has the following problems:
1. It only accepts even length hexadecimal digits (probably because it's made for reading binary): 
    ```python
    >>> unhexlify(b'3fd4c')
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    binascii.Error: Odd-length string
    ```
2. It is indifferent to uppercase/lowercase (both previous solutions accept only lowercase hex strings)
    ```python
    >>> unhexlify(b'3fd4c5') == unhexlify(b'3FD4C5')
    True
    ```

Reply:
1. A hex string is supposed to be even. I don't get your point about this being an issue. Why would we want odd length hexadecimal strings?
2. Why is being indifferent to uppercase/lowercase a problem?

Reply:
I don't know, I assumed that since the `int(string, base)` only accepts lowercase strings to be parsed for base16 conversion, and that the first method only includes lowercase characters, the project's standard is lowercase hexadecimal characters. This isn't a problem with me, I personally want as much fault tolerance as possible.